runtime/vm/BytecodeInterpreter.hpp

```diff
@@ -1898,19 +1901,24 @@ obj:;
 	initialSpecialMethod(REGISTER_ARGS_LIST)
 	{
 		VM_BytecodeAction rc = GOTO_RUN_METHOD;
+		bool split = (JBinvokespecialsplit == _pc[0]);
 		U_16 index = *(U_16*)(_pc + 1);
 		J9ConstantPool *ramConstantPool = J9_CP_FROM_METHOD(_literals);
 		buildGenericSpecialStackFrame(REGISTER_ARGS, 0);
 		updateVMStruct(REGISTER_ARGS);
-		J9Method *method = resolveSpecialMethodRef(_currentThread, ramConstantPool, index, J9_RESOLVE_FLAG_RUNTIME_RESOLVE);
+		if (split) {
+			_sendMethod = resolveSpecialSplitMethodRef(_currentThread, ramConstantPool, index, J9_RESOLVE_FLAG_RUNTIME_RESOLVE);
+		} else {
+			_sendMethod = resolveSpecialMethodRef(_currentThread, ramConstantPool, index, J9_RESOLVE_FLAG_RUNTIME_RESOLVE);
+		}
 		VMStructHasBeenUpdated(REGISTER_ARGS);
 		restoreGenericSpecialStackFrame(REGISTER_ARGS);
 		if (immediateAsyncPending()) {
 			rc = GOTO_ASYNC_CHECK;
 		} else if (VM_VMHelpers::exceptionPending(_currentThread)) {
 			rc = GOTO_THROW_CURRENT_EXCEPTION;
 		} else {
-			_sendMethod = method;
+			rc = invokespecialLogic(REGISTER_ARGS, false, split);
```

Comment:
The order that must be enforced is:

- resolution errors
- NPE
- interface not implemented errors

The initial method is invoked before the NPE and interface checks (logically), so the logic must run again. See invokevirtualLogic.

Reply:
The heart of the "optimization" portion of the change is to run the intial method appropriately instead of doing the NULL method check in invokespecialsplit.